### DISCUSSION TITLE
columnChart - maxY with mergeMode incorrect

### DISCUSSION THREAD

**Discussion by Curwe** [patrykandpatrick/vico#481] [Questions] [answered]
Vico version 1.13.1
Vico core|compose 

Kotlin 1.9.21

########

Hello! 

I have a graph that can show data in two ways. I use mergeMode to change how it's displayed. 

I have 3 buttons below the graph - "Overview", "Production", "Consumption"
Both "Production" and "Consumption" pass MergeMode.Stacked, and "Overview" pass MergeMode.Grouped. 
Overview does show both Production and Consumption data grouped. 

When I press any of the buttons I run some code in the view model to decide what data to show using the ComposedChartEntryModelProducer and the .runTransaction { } function. 

runTransaction has a when check inside that look at what button I pressed and then add the correct data there. 

So here is where my problem start. The startAxis maxY values are wrong when using the MergeMode.Grouped. 
It seem to combine the two Y values per index (x) and then proceed to use the largest number. 
When it really should just take the largest of the two Y values and use that. 

I've tried both without and with a columnChart.axisValuesOverrider and same issue both times. 
My axisValuesOverrider look like this. 
```
val mergeMode = remember(uiState.mergeMode) { uiState.mergeMode }

fun valuesOverrider() =
    object : AxisValuesOverrider<ChartEntryModel> {
        override fun getMaxY(model: ChartEntryModel): Float =
            model.applyOffsetToY { mergeMode }
    } 
```
The _applyOffsetToY_ function looks like this. It just adds a bit extra overhead 
```
fun ChartEntryModel.applyOffsetToY(mergeMode: () -> mergeMode: ColumnChart.MergeMode): Float = with(entries.flatten()) {
    fun Float.offset() = this.plus(if (this < 20) this.plus(1) else this.times(0.1f))

    when (mergeMode()) {
       ColumnChart.MergeMode.Grouped -> maxOfOrNull { it.y }?.coerceAtLeast(0f) ?: 0f
        else -> {
            // When column is pressed we show a more detailed "stack" of information. 
            // This will calculate the Y values for this case. 
            val values = map { it }

            distinctBy { it.x }.map { chartEntry ->
                values.filter { it.x == chartEntry.x }.sumOf { it.y }
            }.maxOfOrNull { it } ?: 0f
        }
    }.offset()
}
```

If I put logs in to the _applyOffsetToY_ function and check the mergeMode used it will always be the previous mergeMode used. So it does not use the currently selected for some reason. I tried to find a way though this as you see using a lambda to always use a fresh mergeMode. But it still use the previous one. 

And as I said above, this is still an issue even if I let the library do my maxY calculations for me. 

I've tried to find a way to come around this issue for some time now, but I'm stuck 🤔 Anyone has any idea?

**Comment by patrickmichalik** [comment, marked as answer]
Hello! The problem is that the `MergeMode` changes and the data updates aren’t in sync. It sounds like each `MergeMode` change occurs after its corresponding data update starts being processed, but before the chart is redrawn for the new data. Generally, in such cases, extras should be used—please see `ComposedChartEntryModelProducer.Transaction#updateExtras`. However, in Vico 1.13, it’s not possible to change a `ColumnChart`’s `MergeMode` based on its `ChartEntryModel`. We’ve introduced such functionality for you in [Vico 2.0.0 Alpha 2](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.2).

For Vico 1.13, a modified version of the workaround you’ve shared may produce better results. Rather than using `uiState.mergeMode` in the `AxisValuesOverrider`, please try adding a `MergeMode` extra to the `ComposedChartEntryModel` and updating the `AxisValuesOverrider` to read the extra. The extra should be updated as `uiState.mergeMode` is. Some synchronization issues remain possible, but the _y_ range should, in general, be correct. Let me know if you need an example.

In Vico 2.0.0 Alpha 2, the custom `AxisValuesOverrider` won’t be necessary. All you’ll need to do is store the `MergeMode` as an extra (rather than in the UI state), an example of which is provided in the release notes, linked above.

**Reply by Curwe** [reply]
Alright I'll try that. Thanks for the tip

**Reply by Curwe** [reply]
Worked perfect! Thanks for that. Helped a lot with the example in the release notes 👍 

```
fun valuesOverrider() = object : AxisValuesOverrider<ChartEntryModel> {
    override fun getMaxY(model: ChartEntryModel): Float = 
        model.applyOffsetToY { model.extraStore[mergeModeKey] }
}
```

Using this together with my AxisValuesOverrider now to add that bit of extra to the Y axis

**Reply by patrickmichalik** [reply]
I’m glad to hear that. You’re welcome!
